### PR TITLE
feat(TextLink): Add TextLink component

### PR DESCRIPTION
**Pull request by kazaky**
## ✅ Checklist

- [x] I have reviewed the submitted code.
- [x] I have tested on a phone device/emulator.
- [x] If it includes design changes, please ask for a review `spark-design` GitHub team.

## 📸 Screenshots
Screenshot 2023-12-11 at 5 02 36 PM

Screenshot 2024-01-05 at 5 58 29 PM

### CODE REVIEW

**Comment by kazaky** [comment]
@spark-ui-bot paparazzi golden images

**Comment by kazaky** [comment]
@spark-ui-bot paparazzi golden images

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import com.adevinta.spark.components.text.TextLink
+import com.adevinta.spark.patchedEnvironment
+import com.adevinta.spark.sparkSnapshot
+import com.android.ide.common.rendering.api.SessionParams
+import com.google.testing.junit.testparameterinjector.TestParameterInjector
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(TestParameterInjector::class)
```
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:46]: It does not seem to be used

File: catalog/src/main/res/drawable-xxhdpi/text_link.png
Comment (SimonMarquis) [inline catalog/src/main/res/drawable-xxhdpi/text_link.png:1]: We should either use a vector drawable or rasters for each density.

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+internal class TextLinkScreenshot {
+
+    private val colorsText: List<Color> = listOf(Color.Blue, Color.Green, Color.Red, Color.Cyan)
+    private val colorsLink: List<Color> = listOf(Color.Magenta, Color.Yellow, Color.Black, Color.LightGray)
```
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:50]: Q: Do we need to "test" this many color combinations?

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        ),
+    )
+
+    @Test
+    fun test() {
+        textToTextLinkPairs.forEach { textToTextLink ->
+            colorsText.forEach { colorText ->
+                colorsLink.forEach { colorLink ->
+                    paparazzi.sparkSnapshot(
+                        name = "Link_${textToTextLink.first.substring(0, 10)}" + "_Color_Link$colorLink",
```
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:82]: nit: I would prefer we use propr name instead of relying on the first 10 chars of test input.

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                        Surface(
+                            color = SparkTheme.colors.surface,
+                        ) {
+                            Row(modifier = Modifier.padding(24.dp)) {
+                                Column(
+                                    verticalArrangement = Arrangement.spacedBy(8.dp),
+                                ) {
+                                    Text(
+                                        text = "Link: ${textToTextLink.first.substring(0, 10)}, ColorLink: $colorLink",
+                                    )
```
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:93]: I don't think adds any useful value to the screenshot test

File: spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_test_link_know_more__color_linkcolor(0.0,_0.0,_0.0,_1.0,_srgb_iec61966-2.1).png
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_test_link_know_more__color_linkcolor(0.0,_0.0,_0.0,_1.0,_srgb_iec61966-2.1).png:1]: It would be great if we merge similar screenshots. Otherwise, if there is a regression, it is most likely that all these will fail the exact same way.

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+```kotlin
+fun TextLink(
+    textFull: String,
+    textLink: String,
+    colorText: Color = Color.Unspecified,
+    colorLink: Color = colorText,
+    onClick: () -> Unit,
+
+    )
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.md:19]:
```suggestion
    onClick: () -> Unit,
)
```

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt (hunk 2 above)
Comment (kazaky) [reply spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:50]: Not necessarily, just adding couple of color combinations together to showcase the component

File: catalog/src/main/res/drawable-xxhdpi/text_link.png
Comment (kazaky) [reply catalog/src/main/res/drawable-xxhdpi/text_link.png:1]: will do

File: spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt (hunk 4 above)
Comment (kazaky) [reply spark-screenshot-testing/src/test/kotlin/com/adevinta/spark/text/TextLinkScreenshot.kt:93]: will remove

File: spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_test_link_know_more__color_linkcolor(0.0,_0.0,_0.0,_1.0,_srgb_iec61966-2.1).png
Comment (kazaky) [reply spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_test_link_know_more__color_linkcolor(0.0,_0.0,_0.0,_1.0,_srgb_iec61966-2.1).png:1]: will delete color combinations then

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    overflow: TextOverflow = TextOverflow.Clip,
+    softWrap: Boolean = true,
+    maxLines: Int = Int.MAX_VALUE,
+    minLines: Int = 1,
+    inlineContent: ImmutableMap<String, InlineTextContent> = persistentMapOf(),
+    onTextLayout: (TextLayoutResult) -> Unit = {},
+    style: TextStyle = LocalTextStyle.current,
+    onClick: () -> Unit,
+) {
+    val start = textFull.indexOf(textLink)
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:144]: This implementation won't work if in the paragraph there's multiple occurence of `textLink` String. It'll always take the first one whereas we can have a first mention of it while the "tappable" part would be at the end of the paragraph for ex.

That's why it should support only a annotatedString and let stylized their paragraph with annotations.

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                fontWeight = FontWeight.Bold,
+                textDecoration = TextDecoration.Underline,
+            ),
+            start,
+            end,
+        )
+    }
+
+    val layoutResult = remember { mutableStateOf<TextLayoutResult?>(null) }
+    val pressIndicator = Modifier.pointerInput(onClick) {
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:161]: pointerInput will return a result when the pointer is exactly on the asked coordinates range and won't be accessible for small text styles.

That's why we wanted to use the clickable modifier on the component to make it fully accessible and support the tap prediction from Android

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ * @param onClick callback when textLink part is clicked.
+ */
+
+@Composable
+public fun TextLink(
+    textFull: String,
+    textLink: String,
+    modifier: Modifier = Modifier,
+    colorText: Color = Color.Unspecified,
+    colorLink: Color = colorText,
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:127]: The link color can also be defined with the string res annotations that we have in Spark that handles our tokens and would allow for a better customization without having for us to support the whole `Text` Api

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    textAlign: TextAlign? = null,
+    lineHeight: TextUnit = TextUnit.Unspecified,
+    overflow: TextOverflow = TextOverflow.Clip,
+    softWrap: Boolean = true,
+    maxLines: Int = Int.MAX_VALUE,
+    minLines: Int = 1,
+    inlineContent: ImmutableMap<String, InlineTextContent> = persistentMapOf(),
+    onTextLayout: (TextLayoutResult) -> Unit = {},
+    style: TextStyle = LocalTextStyle.current,
+    onClick: () -> Unit,
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:142]: The click label which will block the screen reader from giving context to the user tapping on it

**Review by soulcramer — CHANGES_REQUESTED** [review]

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt (hunk 9 above)
Comment (kazaky) [reply spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:142]: @soulcramer I don't get this one, do you suggest something?

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/samples/text/TextLinkExamples.kt [hunk 10]
```diff
@@ -21,69 +21,218 @@
...
+import com.adevinta.spark.components.snackbars.SnackbarHostState
 import com.adevinta.spark.components.text.TextLink
+import com.adevinta.spark.icons.InfoOutline
+import com.adevinta.spark.icons.SparkIcons
+import kotlinx.coroutines.launch
 
 private const val TextLinksExampleSourceUrl = "$SampleSourceUrl/TextLinkExamples.kt"
+
+private val annotatedString1 = buildAnnotatedString {
+    append("Know more about the ")
```
Comment (soulcramer) [inline catalog/src/main/kotlin/com/adevinta/spark/catalog/examples/samples/text/TextLinkExamples.kt:51]: This works but imo it would be better to use the `annotatedStringResource` functions 🤔

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 11]
```diff
@@ -115,80 +106,79 @@ import kotlinx.coroutines.launch
...
-    lineHeight: TextUnit = TextUnit.Unspecified,
+    style: TextStyle = LocalTextStyle.current,
     overflow: TextOverflow = TextOverflow.Clip,
     softWrap: Boolean = true,
     maxLines: Int = Int.MAX_VALUE,
     minLines: Int = 1,
     inlineContent: ImmutableMap<String, InlineTextContent> = persistentMapOf(),
     onTextLayout: (TextLayoutResult) -> Unit = {},
-    style: TextStyle = LocalTextStyle.current,
+    icon: SparkIcon? = null,
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:129]: The icon should not be part of the api on this component, we should let the user do his layout on the `TextLinkButton` when they need an icon or any other component

Comment (kazaky) [reply spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:129]: @soulcramer Can you explain more?
I don't get your point, if you can provide an example it would be nice

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 12]
```diff
@@ -115,80 +106,80 @@ import kotlinx.coroutines.launch
...
 @Composable
 public fun TextLink(
-    textFull: String,
-    textLink: String,
+    text: AnnotatedString,
+    onClickLabel: String,
     modifier: Modifier = Modifier,
-    colorText: Color = Color.Unspecified,
-    colorLink: Color = colorText,
-    fontSize: TextUnit = TextUnit.Unspecified,
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt]: We just need to re-add these arguments

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 13]
```diff
@@ -115,80 +106,80 @@ import kotlinx.coroutines.launch
...
     onTextLayout: (TextLayoutResult) -> Unit = {},
-    style: TextStyle = LocalTextStyle.current,
+    icon: SparkIcon? = null,
+    iconColor: Color = Color.Unspecified,
+    iconSide: IconSide = IconSide.END,
     onClick: () -> Unit,
 ) {
-    val start = textFull.indexOf(textLink)
-    val end = start + textLink.length
+    val layoutResult = remember { mutableStateOf<TextLayoutResult?>(null) }
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:134]: we can remove `layoutResult`

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt [hunk 14]
```diff
@@ -115,80 +106,80 @@ import kotlinx.coroutines.launch
...
+                    .testTag("TextLinkIcon"),
+                contentDescription = null, // Text link text should be enough for context
+            )
+        }
+        SparkText(
+            text = text,
+            modifier = Modifier
+                .weight(1F, false)
+                .alignByBaseline(),
+            style = style.copy(platformStyle = PlatformTextStyle(includeFontPadding = false)),
```
Comment (soulcramer) [inline spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt:158]: We can remove the this as it's applied on every typos on #880

File: spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt (hunk 12 above)
Comment (kazaky) [reply spark/src/main/kotlin/com/adevinta/spark/components/text/TextLink.kt]: Provided a PR, can you kindly have a look
https://github.com/adevinta/spark-android/pull/897

**Review by soulcramer — DISMISSED** [review]

File: spark-screenshot-testing/build.gradle.kts [hunk 15]
```diff
@@ -37,6 +37,7 @@ dependencies {
     implementation(libs.androidx.compose.ui)
     implementation(libs.androidx.compose.material3)
     implementation(libs.kotlinx.collections.immutable)
+    implementation(files("build/intermediates/compile_r_class_jar/debug/R.jar"))
```
Comment (soulcramer) [inline spark-screenshot-testing/build.gradle.kts:40]: Is this intended?

**Review by soulcramer — DISMISSED** [review]

**Comment by kazaky** [comment]
@spark-ui-bot Paparazzi golden images

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/samples/progressbar/ProgressbarConfigurator.kt [hunk 16]
```diff
@@ -88,7 +88,7 @@ private fun ProgressbarSample() {
             isRounded = isRounded,
         )
 
-        Text(text = "ProgressbarIndeterminate")
+        Text(text = "Progressbar Indeterminate")
```
Comment (SimonMarquis) [inline catalog/src/main/kotlin/com/adevinta/spark/catalog/configurator/samples/progressbar/ProgressbarConfigurator.kt:91]: Don't commit unrelated changes in PRs :)

File: catalog/src/main/kotlin/com/adevinta/spark/catalog/model/Components.kt [hunk 17]
```diff
@@ -219,6 +221,19 @@ private val Progressbars = Component(
...
 
+private val TextLinks = Component(
+    id = nextId(),
+    name = "TextLinks",
+    description = R.string.component_textlink_description,
+    illustration = R.drawable.icon_textlink,
+    tintIcon = false,
+    guidelinesUrl = "$ComponentGuidelinesUrl/p/75ed11-textlink/b/403107",
+    docsUrl = "$PackageSummaryUrl/com.adevinta.spark.components.text/index.html",
+    sourceUrl = "$SparkSourceUrl/kotlin/com/adevinta/components/tetx/TextLink.kt",
```
Comment (SimonMarquis) [inline catalog/src/main/kotlin/com/adevinta/spark/catalog/model/Components.kt:232]: This looks like a typo
```suggestion
    sourceUrl = "$SparkSourceUrl/kotlin/com/adevinta/components/text/TextLink.kt",
```

File: catalog/src/main/res/drawable/icon_textlink.xml
Comment (SimonMarquis) [inline catalog/src/main/res/drawable/icon_textlink.xml:1]: The hand really looks weird 😄 
image

File: catalog/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -71,6 +71,8 @@
     <string name="component_progressbar_description">The progress bar component is used to display the length and your progression inside the process or to express a waiting time. This waiting time can be either determinate or indeterminate..</string>
     <string name="component_tag_description">Tags are used to label content and help users quickly recognize info about them: Categories, Status… Can be applied with different colors and designs that are associated with a content due to its characteristics: new content, unvisited content, featured content… Users can’t interact with Tags.</string>
 
+    <string name="component_textlink_description">A textlink is a reference to a resource. It can be external (e.g. a different web page) or internal (e.g. a specific element in the current page).</string>
```
Comment (SimonMarquis) [inline catalog/src/main/res/values/strings.xml:74]: Shouldn't we keep the appropriate case? `TextLink`

File: catalog/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -88,4 +90,11 @@
     <string name="component_menu_issue">Report an issue</string>
 
     <string name="icons_screen_search_helper">Search Spark Icon</string>
-    </resources>
+
+    <string name="spark_text_link_paragraph_example"><annotation typography="body1" color="neutral">This is Adevinta </annotation><annotation typography="body1" color="main"><u><b>Privacy &amp; Policy</b></u></annotation><annotation typography="body1" color="neutral">also lots of extra information you maybe interested in or should I inform you about that extra information. you might be not knowing it</annotation> </string>
```
Comment (SimonMarquis) [inline catalog/src/main/res/values/strings.xml:94]: Trailing space on purpose?
```suggestion
    <string name="spark_text_link_paragraph_example"><annotation typography="body1" color="neutral">This is Adevinta </annotation><annotation typography="body1" color="main"><u><b>Privacy &amp; Policy</b></u></annotation><annotation typography="body1" color="neutral">also lots of extra information you maybe interested in or should I inform you about that extra information. you might be not knowing it</annotation></string>
```

File: catalog/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -88,4 +90,11 @@
...
+
+    <string name="spark_text_link_short_example"><annotation typography="display2" color="neutral">Learn Kotlin Programming </annotation><annotation typography="display3" color="success"><u><b>https://kotlinlang.org</b></u></annotation> </string>
```
Comment (SimonMarquis) [inline catalog/src/main/res/values/strings.xml:96]: same
```suggestion
    <string name="spark_text_link_short_example"><annotation typography="display2" color="neutral">Learn Kotlin Programming </annotation><annotation typography="display3" color="success"><u><b>https://kotlinlang.org</b></u></annotation></string>
```

File: catalog/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -88,4 +90,11 @@
...
+
+    <string name="spark_text_link_short_example_label">https://kotlinlang.org</string>
```
Comment (SimonMarquis) [inline catalog/src/main/res/values/strings.xml:98]: This does not seem to be used?

File: gradle/libs.versions.toml [hunk 22]
```diff
@@ -1,6 +1,7 @@
 [versions]
 accompanist = "0.30.1"
 android-gradlePlugin = "8.2.2"
+colorPicker = "1.0.0"
```
Comment (SimonMarquis) [inline gradle/libs.versions.toml:4]: can we keep the case consistent with the library declaration, either `colorPicker` or `colorpicker`

File: spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_testTextLinkButton_textlinkbutton_accent.png
Comment (SimonMarquis) [inline spark-screenshot-testing/src/test/snapshots/images/com.adevinta.spark.text_TextLinkScreenshot_testTextLinkButton_textlinkbutton_accent.png:1]: Again, we don't need to test all colors I think. This does really bring any value to test every combination of color, size, etc..

File: spark/src/main/res/values-fr/strings.xml [hunk 23]
```diff
@@ -60,4 +60,10 @@
     <string name="spark_textfield_content_description">Obligatoire</string>
     <string name="spark_textfield_delete_a11y">Tout supprimer</string>
 
+    <string name="spark_text_link_paragraph_example"><annotation typography="body1" color="neutral">This is Adevinta </annotation><annotation typography="body1" color="main"><u><b>Privacy &amp; Policy</b></u></annotation><annotation typography="body1" color="neutral">also lots of extra information you maybe interested in or should I inform you about that extra information. you might be not knowing it</annotation> </string>
```
Comment (SimonMarquis) [inline spark/src/main/res/values-fr/strings.xml:63]: I'm not sure why these are added to the public spark package. Shouldn't we only declare them in the catalog or screenshots for testing purposes?

File: catalog/src/main/res/drawable/icon_textlink.xml
Comment (kazaky) [reply catalog/src/main/res/drawable/icon_textlink.xml:1]: Feel free to draw one! 😛

File: gradle/libs.versions.toml (hunk 22 above)
Comment (kazaky) [reply gradle/libs.versions.toml:4]: Alright, I removed this repo actually not required anymore

File: spark/src/main/res/values-fr/strings.xml (hunk 23 above)
Comment (kazaky) [reply spark/src/main/res/values-fr/strings.xml:63]: As it was used in the preview test, I will remove it

File: catalog/src/main/res/values/strings.xml [hunk 24]
```diff
@@ -88,4 +90,9 @@
     <string name="component_menu_issue">Report an issue</string>
 
     <string name="icons_screen_search_helper">Search Spark Icon</string>
-    </resources>
+
+    <string name="spark_text_link_paragraph_example_"><annotation typography="body1" color="neutral">This is Adevinta </annotation><annotation typography="body1" color="main"><u><b>Privacy &amp; Policy</b></u></annotation><annotation typography="body1" color="neutral">also lots of extra information you maybe interested in or should I inform you about that extra information. you might be not knowing it</annotation></string>
```
Comment (SimonMarquis) [inline catalog/src/main/res/values/strings.xml:94]: @kazaky I'm not sure why these have been renamed with trailing underscores 🤔

Comment (kazaky) [reply catalog/src/main/res/values/strings.xml:94]: There was a build error for duplicate res. I just renamed the String res and its fixed
